[13.0][IMP] github_connector: Code analysis (#63)
OCA/interface-github: github_connector/readme/USAGE.rst

   .. image:: ../static/description/github_repository_branch_list.png

Analysis source code
~~~~~~~~~~~~~~~~~~~~~~~~

> It's possible to create custom analaysis rule that can related to Organization and/or Repository to analyze repository code

Reviewer: ```suggestion
It's possible to create custom analysis rules that relate to a GitHub organization, GitHub repository and/or GitHub repository branch to analyze code.
```

#. Go to 'Settings' / 'Analysis rule groups' and create records that allow to organize the rules

#. Go to 'Settings' / 'Analysis rules' and create rules, for example:

> 1. Name: All code, Group: General, Paths: ``*``
> 2. Name: Python code, Group: General, Paths: ``*.py``
> 3. Name: Xml code, Group: General, Paths: ``*.xml``
> 4. Name: Repository 1, Group: Custom, Paths: /path/

Reviewer: Indent this for proper visualization:
```suggestion
   1. Name: All code, Group: General, Paths: ``*``
   2. Name: Python code, Group: General, Paths: ``*.py``
   3. Name: Xml code, Group: General, Paths: ``*.xml``
   4. Name: Repository 1, Group: Custom, Paths: /path/
```

Note: Paths field in 'Analysis rule' allow to put multiple paths for line, path format is according to https://git-scm.com/docs/gitignore#_pattern_format

> #. Go to 'Organizations' and define Analysis rules (optional)
> #. Go to 'Repositories' and define Analysis rules (optional)

Reviewer: Full menu path and include branches:

```suggestion
#. Go to *GitHub > GitHub Community > Organizations* and define Analysis rules (optional)
#. Go to *GitHub > GitHub Repository > Repositories* and define Analysis rules (optional)
#. Go to *GitHub > GitHub Repository > Repository Branches* and define Analysis rules (optional)
```

Reviewer: Now I'm seeing that functionally there's no option of putting rules at branch level, but it should be, as for example folder organization has changed over the branches (versions - OCB 7.0 has /openerp/ folder instead of /odoo)

Author: Maybe this problem can solved with **github_connector_odoo** and `module_paths` field in Repository Branch. https://github.com/OCA/interface-github/pull/64

> Analysis source code is executed when 'Update Source Code Analysis' button in some 'Repository Branch', get all the Analysis rules (Repository + Organization) and analyze code and generate info about it

Reviewer: Also on cron execution

> In 'Repository Branch' / 'Code Analysis' show the info related to Analysis rules

Reviewer: ```suggestion
You can see in 'Repository Branch' / 'Code Analysis' the info obtained from analysis rules.
```

   .. image:: ../static/description/github_repository_branch_list.png

Analysis source code
~~~~~~~~~~~~~~~~~~~~~~~~

It's possible to create custom analaysis rule that can related to Organization and/or Repository to analyze repository code

#. Go to 'Settings' / 'Analysis rule groups' and create records that allow to organize the rules

#. Go to 'Settings' / 'Analysis rules' and create rules, for example:

1. Name: All code, Group: General, Paths: ``*``
2. Name: Python code, Group: General, Paths: ``*.py``
3. Name: Xml code, Group: General, Paths: ``*.xml``
4. Name: Repository 1, Group: Custom, Paths: /path/

> Note: Paths field in 'Analysis rule' allow to put multiple paths for line, path format is according to https://github.com/cpburnz/python-path-specification#tutorial definition

Reviewer: Better to put https://git-scm.com/docs/gitignore#_pattern_format IMO

#. Go to 'Organizations' and define Analysis rules (optional)
#. Go to 'Repositories' and define Analysis rules (optional)

Analysis source code is executed when 'Update Source Code Analysis' button in some 'Repository Branch', get all the Analysis rules (Repository + Organization) and analyze code and generate info about it

> In 'Repository Branh' / 'Code Analysis' show the info related to Analysis rules

Reviewer: ```suggestion
In 'Repository Branch' / 'Code Analysis' show the info related to Analysis rules
```

> Analysis source code is generated by https://github.com/roskakori/pygount

Reviewer: Add in INSTALL.rst the need of pathspec and pygount

Author: Moved this line to `INSTALL.rst` file

**Sizes by Serie**

.. image:: ../static/description/reporting_sizes_by_serie.png

**Repository branch analysis rule**

> .. image:: ../static/description/github_repository_branch_rule_info_report.png

Reviewer: The file has 2 times the extension (`.png.png`).
